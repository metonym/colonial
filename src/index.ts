// TODO: call "rule" "styleObject"

type CSSRules = Partial<CSSStyleDeclaration>;

interface IStyleSheetRule {
  [selector: string]: CSSRules;
}

const toProperty = (fragment: string) => {
  const split = fragment.split('-');

  if (split.length <= 1) {
    return fragment;
  }

  return split
    .map((item, i) => {
      if (i === 0) {
        return item;
      }

      const s = item.split('');
      s[0] = s[0].toUpperCase();

      return s.join('');
    })
    .join('');
};

// TODO: add guaranteed type
interface IObject {
  key: undefined | string;
  rule: {} | IStyleSheetRule;
}

const toObject = (text: string) => {
  const fragments = text.split('\n').filter(i => i.trim().length);

  if (fragments.length === 0) {
    return { key: undefined, rule: {} };
  }

  const cssRules: CSSRules = {};

  text
    .split('\n')
    .filter(i => i.trim().length)
    .forEach(item => {
      const [left, right] = item.split(':');
      const property = toProperty(left).trim();
      const value = right.replace(';', '').trim();
      cssRules[property as any] = value;
    });

  const key = generateKey();
  const obj: IObject = { key, rule: { [key]: cssRules } };

  return obj;
};

interface IRulesByIndex {
  [key: string]: number;
}

const rules: IStyleSheetRule[] = [];
const rulesByIndex: IRulesByIndex = {};

const createRule = (prop: string | IObject) => {
  const obj = typeof prop === 'string' ? toObject(prop) : prop;

  if (obj.key) {
    rules.push(obj.rule);
    rulesByIndex[obj.key] = rules.length;
  }

  return obj;
};

const generateKey = () => {
  // TODO: prefix
  const key = `_${rules.length}`;

  return key;
};

type Css = (args: TemplateStringsArray) => undefined | string;
type Style = (args: CSSRules) => IStyleSheetRule;

const css: Css = args => {
  if (Array.isArray(args)) {
    return createRule(args[0]).key;
  }

  return undefined;
};

const style: Style = args => {
  if (Object.keys(args).length === 0) {
    return {};
  }

  const key = generateKey();
  const obj: IObject = { key, rule: { [key]: args } };

  return createRule(obj).rule;
};

const sheet = () => rules;

interface IColonial {
  css: Css;
  style: Style;
  sheet: () => IStyleSheetRule[];
}

const colonial: IColonial = { css, style, sheet };

export default colonial;
export { css, style, sheet };
