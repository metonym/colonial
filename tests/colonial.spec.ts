import { css, sheet, style } from '../src';

describe('colonial', () => {
  test('css, sheet, style', () => {
    // @ts-ignore
    expect(css()).toBeUndefined();

    // @ts-ignore
    expect(style({})).toEqual({});

    expect(
      css`
        color: red;
      `
    ).toEqual(expect.any(String));
    expect(sheet().length).toEqual(1);

    expect(
      css`
        color: red;
        background-color: blue;
      `
    ).toEqual(expect.any(String));
    expect(sheet().length).toEqual(2);

    expect(
      style({
        color: 'red',
        backgroundColor: 'blue'
      })
    ).toEqual(expect.any(Object));
    expect(sheet().length).toEqual(3);

    expect(css``).toBeUndefined();
    expect(sheet().length).toEqual(3);
    expect(sheet()).toMatchSnapshot();
  });
});
