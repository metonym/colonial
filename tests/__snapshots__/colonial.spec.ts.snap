// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colonial css, sheet, style 1`] = `
Array [
  Object {
    "_0": Object {
      "color": "red",
    },
  },
  Object {
    "_1": Object {
      "backgroundColor": "blue",
      "borderBottomWidth": "4px",
      "color": "red",
    },
  },
  Object {
    "_2": Object {
      "backgroundColor": "blue",
      "color": "red",
    },
  },
]
`;
